src/IECoreHoudini/SOP_SceneCacheSource.cpp

```diff
@@ -231,6 +239,13 @@ OP_ERROR SOP_SceneCacheSource::cookMySop( OP_Context &context )
 	double readTime = time( context );
 	Imath::M44d transform = ( space == World ) ? worldTransform( file, path, readTime ) : Imath::M44d();
 
+	bool foundInvisible;
+	invisibleLocations( scene.get(), foundInvisible, readTime );
+	if( foundInvisible )
+	{
+		gdp->clearAndDestroy();
```

Comment:
If I understand correctly, this means that we will always destroy everything and rebuild from scratch, any time there is an invisible location? And we'll be doing that regardless of the setting for `pVisibilityFilter`, which may be off anyway? I'm not familiar with this code, but there are comments below that suggest that we can rebuild lazily in the animated case, so forcing a rebuild from scratch seems particularly undesirable.

Is it perhaps sufficient to just remove this check and instead include the value of `pVisibilityFilter` in the hash above? Then we'd only rebuild from scratch when the parameter value has changed.

Reply:
there are multiple things here so I'll try to answer each on of them:
- I believe the test for `pVisibilityFilter` parameter was already happening but inside `invisibleLocations`, I took it out so it's a bit more clear what's happening.
- I'm also new to this code but as far as I can tell, Houdini uses the previous cook as a starting point unless you are running `clearAndDestroy`, in the animated case we are simply updating prim vars. I'm not too sure how this handle changing topology...
- I did try to add `pVisibilityFilter` in the hash at first but I think this is unsufficient, what we would need is that modification of the visibility attribute in any location to contribute to the hash. Which is essentially what I'm doing by calling `clearAndDestroy` directly.

I hope that makes sense, let me know if you have any questions.

Reply:
> I believe the test for pVisibilityFilter parameter was already happening but inside invisibleLocations, I took it out so it's a bit more clear what's happening.

Thanks - that is clearer.

> I'm also new to this code but as far as I can tell, Houdini uses the previous cook as a starting point unless you are running clearAndDestroy, in the animated case we are simply updating prim vars. 

That is my (limited) understanding too. My concern is that with this change we have totally lost this optimisation for any scene with visibility attributes. Not only have we lost it, but we've added an extra traversal of the _entire_ scene cache. The object loading code seems to short-cut for locations we don't want to load, and we're not even doing that.

> I did try to add pVisibilityFilter in the hash at first but I think this is unsufficient, what we would need is that modification of the visibility attribute in any location to contribute to the hash.

If I understand correctly, I'm not even sure that `if invisibleLocations()` is sufficient. What if the previous cook had one invisible location, and the current cook has none (because of animated visibility). Won't that mean we fail to load the newly-visible object? In an ideal world, I think `loadObjects()` would deal with visibility, selectively deleting things it doesn't need and adding things it does. Is that possible?

Like I say, I'm not very familiar with this code, but given my current level of understanding, I don't feel able to sign off on it either from a performance or correctness standpoint.

Reply:
Your concern makes sense.

As far as I know, `clearAndDestroy` is deleting the entire flat list of geometry so doing is at part of the recursive `loadObjects()` would mean deleting whatever we have already loaded that's why I'm doing this as a pre-pass before loading anything.

Indeed we are loosing that optimisation but I don't see any other way in that case. No matter what it's true my traversal could be a bit optimized like the `loadObjects`.

`invisibleLocations` is only concerned about destroying geometry previously loaded, so if a location was hidden at the previous time step, it won't affect the loading of that location at the current time step.

I was able to test that with a production scene cache with visible location turning invisble and invisible one turning visible.

I will have a look if HDK has a way to delete geometry selectively, maybe that could be used inside `loadObjects` removing the extra traversal.